### PR TITLE
[FEATURE] Compiler for OGR expressions

### PR DESCRIPTION
**Pull request by nyalldawson**
Implements an expression compiler for OGR sources, which allows much faster retrieval of features when a FilterExpression feature request is used.

OGR SQL supports only basic expressions, so this is really only enabled for "attribute"='value' type expressions. OGR SQL also does case-insensitive matching of strings, so if an expression uses string comparison then we only return a "Partial" compilation success and need to double check the results using QGIS' expression engine (still a huge performance win regardless, especially if there's an index on the OGR data source).

### CODE REVIEW

**Comment by nirvn** [comment]
Super exciting stuff.

Any chance of having complier for spatialite too?

**Comment by nyalldawson** [comment]
@nirvn mmm.... I basically never use spatialite, so will leave that to someone who's motivated ;)

**Comment by nirvn** [comment]
@nyalldawson, hah, fair enough :smile:

**Comment by m-kuhn** [comment]
Does OGR support three valued logic? Do you know what OGR does with the expression internally? Does it forward them to the databases modified/unmodified or is it used only for files with indexes?

**Comment by m-kuhn** [comment]
Hmmm... Should we add a new class QgsSqlExpressionCompiler which does the trivial operator stuff which is the same for all SQL providers? Inheriting classes can add their own additional functions (like spatial functions which tend to be provider or even version specific) or overwrite certain operators if there is the need to?

**Comment by nyalldawson** [comment]
> Does OGR support three valued logic?

Not sure... at least partially it does. I'll add tests for this to confirm.

> Do you know what OGR does with the expression internally? Does it forward them to the databases modified/unmodified or is it used only for files with indexes?

According to http://www.gdal.org/ogr_sql.html:
"all OGR drivers for database systems: MySQL, PostgreSQL and PostGIS (PG), Oracle (OCI), SQLite, ODBC, ESRI Personal Geodatabase (PGeo) and MS SQL Spatial (MSSQLSpatial) override the GDALDataset::ExecuteSQL() function with dedicated implementation and, by default, pass the SQL statements directly to the underlying RDBMS"

Possibly we should test for and disable the expression compilation if the datasource is one of these formats.

> Hmmm... Should we add a new class QgsSqlExpressionCompiler which does the trivial operator stuff which is the same for all SQL providers? Inheriting classes can add their own additional functions (like spatial functions which tend to be provider or even version specific) or overwrite certain operators if there is the need to?

While the general structure of the class is similar to QgsPostgresExpressionCompiler, it's subtly different throughout due to the handling of OGR's case insensitive matching. I'm not sure it's worth the effort for generalising this to a new class, as I expect every provider to have a bunch of subtle differences like this we'd need to handle...

**Comment by m-kuhn** [comment]
On 11/16/2015 08:14 AM, Nyall Dawson wrote:
>
>     Do you know what OGR does with the expression internally? Does it
>     forward them to the databases modified/unmodified or is it used
>     only for files with indexes?
>
> According to http://www.gdal.org/ogr_sql.html:
> "all OGR drivers for database systems: MySQL, PostgreSQL and PostGIS
> (PG), Oracle (OCI), SQLite, ODBC, ESRI Personal Geodatabase (PGeo) and
> MS SQL Spatial (MSSQLSpatial) override the GDALDataset::ExecuteSQL()
> function with dedicated implementation and, by default, pass the SQL
> statements directly to the underlying RDBMS"
>
> Possibly we should test for and disable the expression compilation if
> the datasource is one of these formats.
>
Sounds good to me.
>
>     Hmmm... Should we add a new class QgsSqlExpressionCompiler which
>     does the trivial operator stuff which is the same for all SQL
>     providers? Inheriting classes can add their own additional
>     functions (like spatial functions which tend to be provider or
>     even version specific) or overwrite certain operators if there is
>     the need to?
>
> While the general structure of the class is similar to
> QgsPostgresExpressionCompiler, it's subtly different throughout due to
> the handling of OGR's case insensitive matching. I'm not sure it's
> worth the effort for generalising this to a new class, as I expect
> every provider to have a bunch of subtle differences like this we'd
> need to handle...
>

I think we could still add the base class and handle the subtle
differences in the subclass (More than 90% is just duplicated copy-paste
code). As long as we have everything covered with tests and make sure
that for every provider all the tests are run we should quickly see for
which providers we need custom adjustements and if fixes need to be
applied to the base class or provider specific.

-- 
Matthias Kuhn
OPENGIS.ch - https://www.opengis.ch
Spatial • (Q)GIS • PostGIS • Open Source

**Comment by nyalldawson** [comment]
>> Does OGR support three valued logic?
> Not sure... at least partially it does. I'll add tests for this to confirm.

Done

>According to http://www.gdal.org/ogr_sql.html:
"all OGR drivers for database systems: MySQL, PostgreSQL and PostGIS (PG), Oracle (OCI), SQLite, ODBC, ESRI Personal Geodatabase (PGeo) and MS SQL Spatial (MSSQLSpatial) override the GDALDataset::ExecuteSQL() function with dedicated implementation and, by default, pass the SQL statements directly to the underlying RDBMS"

>Possibly we should test for and disable the expression compilation if the datasource is one of these formats.

Done

**Comment by nyalldawson** [comment]
@m-kuhn Ok, refactored to have a common base class.

@nirvn you're welcome ;)

**Comment by nirvn** [comment]
@nyalldawson wouhou, thanks.

I've just tested your branch, and the results are very good. On a large dataset (1.6 million points):
- selecting via expression against shapefile: master 16.13sec vs. this PR 11.91sec
- selecting via expression against spatialite: master 12.05sec vs. this PR *4.44sec*

Very yay!

**Comment by nirvn** [comment]
For spatialite, it goes all the way down to 3.18sec against an indexed field.

**Comment by m-kuhn** [comment]
Thanks a lot @nyalldawson

Any idea why the tests fail? Is it the same reason that master yesterday was failing for some hours?

**Comment by nyalldawson** [comment]
@m-kuhn rebased against master, Travis is all happy again

File: src/core/qgssqlexpressioncompiler.h [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+class CORE_EXPORT QgsSqlExpressionCompiler
+{
+  public:
+
+    /** Possible results from expression compilation */
+    enum Result
+    {
+      None, /*!< No expression */
+      Complete, /*!< Expression was successfully compiled and can be completed delegated to provider */
```
Comment (m-kuhn) [inline src/core/qgssqlexpressioncompiler.h]: completely

**Comment by nyalldawson** [comment]
@m-kuhn ok, fixed that. I've just added some new tests & fixes for OGR too. 

Additionally, I've replaced the postgres- specific option with a general "enable expression compilation" option, and made it enabled by default (we can change this pre 2.14 release if testing proves poor).

**Comment by m-kuhn** [comment]
Perfect, thanks a lot Nyall!!!

**Pull request merged**